presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java

```diff
@@ -210,8 +210,8 @@
         ProcessorState(Type type, boolean needsMoreData, Optional<T> result, Optional<ListenableFuture<?>> blocked)
         {
             checkArgument(!needsMoreData || type == NEEDS_MORE_DATA || type == RESULT);
-            checkArgument(!blocked.isPresent() || type == BLOCKED);
-            checkArgument(!result.isPresent() || type == RESULT);
+            checkArgument(blocked.isPresent() == (type == BLOCKED));
```

Comment:
nit: I would leave the previous code as conditions are more streamlined

Reply:
I am actually _changing_ the semantics here. Previously you could have `blocked.isPresent()` in any state. Now you can have `blocked.isPresent()` only in `BLOCKED` state.